motionplan/inverseKinematics.go

```diff
@@ -2,23 +2,182 @@ package motionplan
 
 import (
 	"context"
+	"encoding/json"
+	"sort"
+
+	"github.com/edaniels/golog"
+	"go.viam.com/utils"
 
 	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
 	spatial "go.viam.com/rdk/spatialmath"
 )
 
 // InverseKinematics defines an interface which, provided with a goal position and seed inputs, will output all found
 // solutions to the provided channel until cancelled or otherwise completes.
-type InverseKinematics interface {
-	// Solve receives a context, the goal arm position, and current joint angles.
-	Solve(context.Context, chan<- []referenceframe.Input, spatial.Pose, []referenceframe.Input, Metric, int) error
+type InverseKinematicsSolver interface {
+	solve(context.Context, chan<- []referenceframe.Input, spatial.Pose, []referenceframe.Input, Metric, int) error
+	frame() referenceframe.Frame
+	options() *ikOptions
+}
+
+type ikSolver struct {
+	model  referenceframe.Frame
+	logger golog.Logger
+	opts   *ikOptions
+}
+
+func (ik *ikSolver) frame() referenceframe.Frame {
+	return ik.model
 }
 
-func limitsToArrays(limits []referenceframe.Limit) ([]float64, []float64) {
-	var min, max []float64
-	for _, limit := range limits {
-		min = append(min, limit.Min)
-		max = append(max, limit.Max)
+func (ik *ikSolver) options() *ikOptions {
+	return ik.opts
+}
+
+func NewIKSolver(frame referenceframe.Frame, logger golog.Logger, ikConfig map[string]interface{}) (InverseKinematicsSolver, error) {
+	// Start with normal options
+	opt := newBasicIKOptions()
+	opt.extra = ikConfig
+
+	// convert map to json, then to a struct, overwriting present defaults
+	jsonString, err := json.Marshal(ikConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(jsonString, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	// infer IK solver to build based on number of threads allowed
+	if opt.NumThreads <= 1 {
+		return newNLOptIKSolver(frame, logger, opt)
+	}
+	return newEnsembleIKSolver(frame, logger, opt)
+}
+
+func BestIKSolutions(
+	ctx context.Context,
+	ik InverseKinematicsSolver,
+	goal spatialmath.Pose,
+	input []referenceframe.Input,
+	worldState *referenceframe.WorldState,
+	randseed int,
```

Comment:
randseed should either use a default, or be taken from IK options

Reply:
I think my preferred way to do this is: put `randseed` in the `ikConfig` generic map input, and use a default of 1 if the user doesn't pass anything in. Since IK behavior is defined not only by the random seed but also by the seed input, this should work fine.

Motion planning should call `getSolutions` anyway rather than this.